### PR TITLE
TLS-Client - use easy-connect

### PR DESCRIPTION
**Pull request by JanneKiiskila**
Take easy-connect into use with the tls-client, so that one can
more easily (i.e. w/o modifying the actual main.cpp) use multiple
network stacks (just modify the mbed_app.json).

### CODE REVIEW

**Comment by JanneKiiskila** [comment]
Fix for issue; https://github.com/ARMmbed/mbed-os-example-tls/issues/100

**Comment by JanneKiiskila** [comment]
@adbridge  @0xc0170

**Comment by JanneKiiskila** [comment]
rebuild

**Comment by JanneKiiskila** [comment]
The test is failing, because it cant' build. It can't build, because it does not seem to do mbed deploy for it dependencies -> easy connect is not being pulled in. K64F for example fails due to that. Why isn't the build script doing `mbed deploy`? Speed? It's a slow operation...
@adbridge @0xc0170

**Comment by adbridge** [comment]
@mazimkhan

**Comment by mazimkhan** [comment]
We don't do ```mbed deploy``` as it needs git credentials on the slave machine. We rather check the mbed-os commit SHA from mbed-os.lib file and checkout the required version.

**Comment by mazimkhan** [comment]
I see you have added easy-connect.lib that requires ```mbed deploy```. I will try to update the job to do mbed deploy.

**Comment by JanneKiiskila** [comment]
Alternatively you can pull in the easy-connect to the work area, so that it's readily available.

**Comment by andresag01** [comment]
@JanneKiiskila: Thanks for the proposal, we will review it soon.

**Comment by JanneKiiskila** [comment]
@andresag01  - any update?

**Comment by JanneKiiskila** [comment]
easy-connect.lib hash would already have to be updated a bit... WIFI_REALTEK has been changed to WIFI_RTW to better match the rest of the WIFI stacks.

**Comment by JanneKiiskila** [comment]
Updates done.

**Comment by andresag01** [comment]
@JanneKiiskila: I have modified our CI so that it does `mbed deploy` before trying to compile the example. However, we noticed that the example fails to build when using the changes in this PR. We are getting the following error:
```
Compile [  0.2%]: NanostackRfPhyAtmel.cpp
[Fatal Error] NanostackRfPhyAtmel.cpp@18,44: nanostack/platform/arm_hal_phy.h: No such file or directory
[ERROR] ./easy-connect/atmel-rf-driver/source/NanostackRfPhyAtmel.cpp:18:44: fatal error: nanostack/platform/arm_hal_phy.h: No such file or directory
 #include "nanostack/platform/arm_hal_phy.h"
                                            ^
compilation terminated.
```
According to the documentation [here](https://github.com/ARMmbed/easy-connect/blob/master/README.md) we need to exclude nanostack compilation:
>If you encounter a compilation error such as below, you need to add an .mbedignore file that tells the mbed compiler to skip compiling the files that require Nanostack. By default, the mbed compiler compiles every single file from all folders.

Could you please add a suitable `.mbedignore` to the tls-client example? Thanks!

**Comment by JanneKiiskila** [comment]
@andresag01  - roger that, the additional commit will fix that. There are 2 ways to resolve this, 
1) either via `.mbedignore`  OR
2) add the feature `nanostack` to enable compilation.

I've been earlier forced to avoid 2) because I had issues with the compiler sometimes producing a binary, that did not connect to absolutely anything. Latest trials are again working, so I'd prefer that now - it's a bit more end-user friendly.

**Comment by JanneKiiskila** [comment]
Tests failed again, but this time to some Hudson/Java issue... 
```
etting pull request status Cam-CI uvisor example Build & Test to FAILURE with message: Failed
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
java.io.IOException: remote file operation failed: c:\jsbld\ws\mbed-os\e_g_pr_tls@13 at hudson.remoting.Channel@6eb5794:e104819_builds: java.nio.file.FileSystemException: c:\jsbld\ws\mbed-os\e_g_pr_tls@13\tls-client@tmp\durable-c9807502\jenkins-log.txt: The process cannot access the file because it is being used by another process.

	at hudson.FilePath.act(FilePath.java:986)
	at hudson.FilePath.act(FilePath.java:968)
	at hudson.FilePath.deleteRecursive(FilePath.java:1170)
	at org.jenkinsci.plugins.workflow.steps.DeleteDirStep$Execution.run(DeleteDirStep.java:68)
	at org.jenkinsci.plugins.workflow.steps.DeleteDirStep$Execution.run(DeleteDirStep.java:61)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:52)
	at hudson.security.ACL.impersonate(ACL.java:213)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:49)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	Suppressed: java.io.IOException: remote file operation failed: c:\jsbld\ws\mbed-os\e_g_pr_tls@13 at hudson.remoting.Channel@6eb5794:e104819_builds: java.nio.file.FileSystemException: c:\jsbld\ws\mbed-os\e_g_pr_tls@13\tls-client@tmp\durable-c9807502\jenkins-log.txt: The process cannot access the file because it is being used by another process.

		... 13 more
	Caused by: java.nio.file.FileSystemException: c:\jsbld\ws\mbed-os\e_g_pr_tls@13\tls-client@tmp\durable-c9807502\jenkins-log.txt: The process cannot access the file because it is being used by another process.
```

**Comment by andresag01** [comment]
@JanneKiiskila: Thanks for reworking your PR.

Where did you observe this issue? It seems that the tests run, but there were a few problems with the TLS client:
* I think this PR changes the baud rate for the tls-client from 9600 to 115200. Please revert this change to the default as all our infrastructure and the other sample applications are set up to use 9600. Otherwise the tests will not pass.
* I am observing a compilation problem when trying to build tls-client with ARMCC due to a number of undefined symbols. The full error is logged [here](http://e108747.cambridge.arm.com:8080/job/mbed-os/job/e_g_pr_tls/216/execution/node/1184/log/).

Other than this, it seems that the tests should pass with IAR and GCC_ARM. I will soon start reviewing the code itself.

Thanks!

File: tls-client/mbed_app.json [hunk 1]
```diff
@@ -2,8 +2,36 @@
...
+        }
+    },
     "target_overrides": {
-        "UBLOX_EVK_ODIN_W2": {
+        "*": {
+            "target.features_add": ["NANOSTACK"],
+            "platform.stdio-baud-rate": 115200,
+            "platform.stdio-convert-newlines": true
+        },
+            "UBLOX_EVK_ODIN_W2": {
```
Comment (andresag01) [inline tls-client/mbed_app.json]: Please fix the indentation in this case

File: tls-client/mbed_app.json [hunk 2]
```diff
@@ -2,8 +2,36 @@
...
+        },
+        "wifi-password": {
+            "value": "\"Password\""
+        }
+    },
     "target_overrides": {
-        "UBLOX_EVK_ODIN_W2": {
+        "*": {
+            "target.features_add": ["NANOSTACK"],
+            "platform.stdio-baud-rate": 115200,
```
Comment (andresag01) [inline tls-client/mbed_app.json]: Please note that the default is 9600 and all our infrastructure (and examples) are set up to listen at 9600. Please revert this change, otherwise the tests will not pass.

File: tls-client/mbed_app.json [hunk 3]
```diff
@@ -2,8 +2,36 @@
...
+            "value": "\"SSID\""
+        },
+        "wifi-password": {
+            "value": "\"Password\""
+        }
+    },
     "target_overrides": {
-        "UBLOX_EVK_ODIN_W2": {
+        "*": {
+            "target.features_add": ["NANOSTACK"],
```
Comment (andresag01) [inline tls-client/mbed_app.json]: As mentioned before, this is causing build failures with ARMCC

**Comment by mazimkhan** [comment]
retest CI

File: tls-client/mbed_app.json (hunk 3 above)
Comment (JanneKiiskila) [reply tls-client/mbed_app.json]: Then we have no other option than use the `.mbedignore`.

File: tls-client/main.cpp [hunk 4]
```diff
@@ -418,18 +419,15 @@ int main() {
...
-    eth_iface.connect();
-    mbedtls_printf("Using Ethernet LWIP\r\n");
-    const char *ip_addr = eth_iface.get_ip_address();
-    if (ip_addr) {
-        mbedtls_printf("Client IP Address is %s\r\n", ip_addr);
-    } else {
-        mbedtls_printf("No Client IP Address\r\n");
+    /* Use the easy-connect lib to support multiple network bearers.   */
+    /* See https://github.com/ARMmbed/easy-connect README.md for info. */
+    NetworkInterface* network = easy_connect(true); /* has 1 argument, enable_logging (pass in true to log to serial port) */
```
[the remark this answers is not in the document]
Comment (JanneKiiskila) [reply tls-client/main.cpp]: Makes sense.

File: tls-client/mbed_app.json (hunk 2 above)
Comment (JanneKiiskila) [reply tls-client/mbed_app.json]: I think these tests would run faster with 115200? That's the quite standard baud rate for anything these days and all boards I'm aware of (admittedly that's not 100% of boards) - they all can do 115200. How difficult would it be to update the test infra?

But, I do get your point and updating the test infra + this change should be coordinated together.

**Comment by JanneKiiskila** [comment]
Changes done per the review comments, re-based against the repo master and force pushed in.

File: tls-client/main.cpp [hunk 5]
```diff
@@ -418,18 +419,15 @@ int main() {
...
+    if (!network) {
+        printf("Connecting to the network failed... See serial output.\r\n");
+        return 1;
     }
 
-    HelloHTTPS *hello = new HelloHTTPS(HTTPS_SERVER_NAME, HTTPS_SERVER_PORT, &eth_iface);
+    HelloHTTPS *hello = new HelloHTTPS(HTTPS_SERVER_NAME, HTTPS_SERVER_PORT, network);
     hello->startTest(HTTPS_PATH);
```
Comment (k-stachowiak) [inline tls-client/main.cpp]: Pedantic note: if the method startTest throws, the hello object won't be released.

Comment (JanneKiiskila) [reply tls-client/main.cpp]: I think the throwing doesn't work in any embedded system, if something get's thrown - board resets?

**Comment by JanneKiiskila** [comment]
This one needs again heavy rebasing and addition of that `.mbedignore` file, there's a newer version of easy-connect that brings yet another driver that needs a new line in the `.mbedignore`. I'm running out of time now though.

File: tls-client/main.cpp (hunk 5 above)
Comment (TeroJaasko) [reply tls-client/main.cpp]: C++ exceptions are disabled on mbed-os build profiles, https://github.com/ARMmbed/mbed-os/blob/master/tools/profiles/develop.json#L5

**Comment by JanneKiiskila** [comment]
Did a major squash operation to decrease commits & force pushed it in.

**Comment by JanneKiiskila** [comment]
There seems to be some CI issue? 

```
Failed in branch build-authcrypt-K64F-ARM
```
and that seems to fail because: 

```
[build-authcrypt-K64F-ARM] Cloning repository git@github.com:ARMmbed/mbed-os-example-tls.git
[build-authcrypt-K64F-ARM] ERROR: Failed to clean the workspace
[build-authcrypt-K64F-ARM] java.nio.file.FileSystemException: c:\jsbld\ws\mbed-os\e_g_pr_tls@4\tls-client@tmp\durable-19ffe8a3\jenkins-log.txt: The process cannot access the file because it is being used by another process.
[build-authcrypt-K64F-ARM] 
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.nio.file.Files.delete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[build-authcrypt-K64F-ARM] 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.lang.reflect.Method.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteFile(Util.java:257)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:320)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:311)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:311)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:454)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:152)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Request$2.run(Request.java:332)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.FutureTask.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Engine$1$1.run(Engine.java:85)
[build-authcrypt-K64F-ARM] 	at java.lang.Thread.run(Unknown Source)
[build-authcrypt-K64F-ARM] ERROR: Error cloning remote repo 'origin'
[build-authcrypt-K64F-ARM] hudson.plugins.git.GitException: Failed to delete workspace
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:457)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:152)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Request$2.run(Request.java:332)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.FutureTask.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Engine$1$1.run(Engine.java:85)
[build-authcrypt-K64F-ARM] 	at java.lang.Thread.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at ......remote call to e104819_builds(Native Method)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserResponse.retrieve(UserRequest.java:252)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Channel.call(Channel.java:781)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
[build-authcrypt-K64F-ARM] 	at sun.reflect.GeneratedMethodAccessor389.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[build-authcrypt-K64F-ARM] 	at java.lang.reflect.Method.invoke(Method.java:606)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
[build-authcrypt-K64F-ARM] 	at com.sun.proxy.$Proxy66.execute(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1046)
[build-authcrypt-K64F-ARM] 	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1086)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:83)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:73)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:52)
[build-authcrypt-K64F-ARM] 	at hudson.security.ACL.impersonate(ACL.java:213)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:49)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[build-authcrypt-K64F-ARM] 	at java.lang.Thread.run(Thread.java:745)
[build-authcrypt-K64F-ARM] Caused by: java.nio.file.FileSystemException: c:\jsbld\ws\mbed-os\e_g_pr_tls@4\tls-client@tmp\durable-19ffe8a3\jenkins-log.txt: The process cannot access the file because it is being used by another process.
[build-authcrypt-K64F-ARM] 
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.nio.file.Files.delete(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[build-authcrypt-K64F-ARM] 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.lang.reflect.Method.invoke(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteFile(Util.java:257)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:320)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:311)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteRecursive(Util.java:311)
[build-authcrypt-K64F-ARM] 	at hudson.Util.deleteContentsRecursive(Util.java:222)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:454)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
[build-authcrypt-K64F-ARM] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:152)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Request$2.run(Request.java:332)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.FutureTask.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[build-authcrypt-K64F-ARM] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[build-authcrypt-K64F-ARM] 	at hudson.remoting.Engine$1$1.run(Engine.java:85)
[build-authcrypt-K64F-ARM] 	at java.lang.Thread.run(Unknown Source)
```

**Comment by mazimkhan** [comment]
@JanneKiiskila I have fixed the issue and restarted the CI.

**Comment by JanneKiiskila** [comment]
`Failed in branch build-tls-client-NUMAKER_PFM_NUC472-GCC_ARM` - something failed there, seems it can't get connection up. Do we know this HW is OK in this test suite?

**Comment by JanneKiiskila** [comment]
Alternative Oulu CI also seems to have the same deploy issue, it's not pulling in the easy-connect repo.

```
[build-tls-client-K64F-ARM] [Error] main.cpp@36,0:  #5: cannot open source input file "easy-connect.h": No such file or directory
09:22:15 [build-tls-client-K64F-ARM] [ERROR] "./main.cpp", line 36: Error:  #5: cannot open source input file "easy-connect.h": No such file or directory
```

**Comment by mazimkhan** [comment]
I can see tls-client example failing on Numaker and Nucleo boards. But it passes on K64F and Ublox. This test and these platforms are known to be flaky in CI. Alternative CI is more of staging setup before we switch to it.
Given the flakiness of the test and these boards I would consider it pass.

**Comment by JanneKiiskila** [comment]
CI problem; NUVOTON NUMAKER PFM-NUC472 passes when testing manually.

```
TLS connection to os.mbed.com established
Server certificate:
    cert. version     : 3
    serial number     : 65:7B:6D:8D:15:A5:B6:86:87:6B:5E:BC
    issuer name       : C=BE, O=GlobalSign nv-sa, CN=GlobalSign Organization Val                                                           idation CA - SHA256 - G2
    subject name      : C=GB, ST=Cambridgeshire, L=Cambridge, O=ARM Ltd, CN=*.mb                                                           ed.com
    issued  on        : 2017-04-03 13:54:02
    expires on        : 2018-05-06 10:31:02
    signed using      : RSA with SHA-256
    RSA key size      : 2048 bits
    basic constraints : CA=false
    subject alt name  : *.mbed.com, mbed.org, *.mbed.org, mbed.com
    key usage         : Digital Signature, Key Encipherment
    ext key usage     : TLS Web Server Authentication, TLS Web Client Authentica                                                           tion
Certificate verification passed

HTTPS: Received 440 chars from server
HTTPS: Received 200 OK status ... [OK]
HTTPS: Received 'Hello world!' status ... [OK]
HTTPS: Received message:

HTTP/1.1 200 OK
Server: nginx/1.11.10
Date: Fri, 13 Oct 2017 06:22:55 GMT
Content-Type: text/plain
Content-Length: 14
Connection: keep-alive
Last-Modified: Fri, 27 Jul 2012 13:30:34 GMT
Accept-Ranges: bytes
Cache-Control: max-age=36000
Expires: Fri, 13 Oct 2017 16:22:55 GMT
X-Upstream-L3: 172.17.0.4:80
X-Upstream-L2: developer-sjc-indigo-2-nginx
Strict-Transport-Security: max-age=31536000; includeSubdomains

Hello world!
```

**Comment by JanneKiiskila** [comment]
Manual testing shows it works also with UBLOX_ODIN_EVK_W2, but it requires one to use Mbed OS 5.5.2.

**Comment by JanneKiiskila** [comment]
UBLOX works even with Mbed OS 5.6.2. Seems I somehow managed to jam the device nbr 6, but number 5 works just fine. 

Added a few printouts to the beginning, helps to diagnose issues with CI etc, if you get a solid print from the start before ANYTHING else is done. Added also Mbed OS version print.

**Comment by JanneKiiskila** [comment]
Printout from the UBLOX ODIN EVK W2.

```
$$$12360221014F67073E0BE40C
Starting mbed-os-example-tls/tls-client
Using Mbed OS 5.6.2
[EasyConnect] IPv4 mode
Connecting with os.mbed.com
Starting the TLS handshake...
TLS connection to os.mbed.com established
Server certificate:
cert. version : 3
serial number : 65:7B:6D:8D:15:A5:B6:86:87:6B:5E:BC
issuer name : C=BE, O=GlobalSign nv-sa, CN=GlobalSign Organization Validation CA - SHA256 - G2
subject name : C=GB, ST=Cambridgeshire, L=Cambridge, O=ARM Ltd, CN=*.mbed.com
issued on : 2017-04-03 13:54:02
expires on : 2018-05-06 10:31:02
signed using : RSA with SHA-256
RSA key size : 2048 bits
basic constraints : CA=false
subject alt name : *.mbed.com, mbed.org, *.mbed.org, mbed.com
key usage : Digital Signature, Key Encipherment
ext key usage : TLS Web Server Authentication, TLS Web Client Authentication
Certificate verification passed

HTTPS: Received 440 chars from server
HTTPS: Received 200 OK status ... [OK]
HTTPS: Received 'Hello world!' status ... [OK]
HTTPS: Received message:

HTTP/1.1 200 OK
Server: nginx/1.11.10
Date: Sat, 14 Oct 2017 18:58:31 GMT
Content-Type: text/plain
Content-Length: 14
Connection: keep-alive
Last-Modified: Fri, 27 Jul 2012 13:30:34 GMT
Accept-Ranges: bytes
Cache-Control: max-age=36000
Expires: Sun, 15 Oct 2017 04:58:31 GMT
X-Upstream-L3: 172.17.0.4:80
X-Upstream-L2: developer-sjc-indigo-1-nginx
Strict-Transport-Security: max-age=31536000; includeSubdomains

Hello world!
```

**Comment by JanneKiiskila** [comment]
I don't think anything should block the merge anymore, right? 
- NUMAKER and UBLOX verified manually that they pass.

**Comment by AndrzejKurek** [comment]
Everything looks fine for me.

**Comment by andresag01** [comment]
@k-stachowiak: It seems that there have been replies to your comments. Could you please review this PR again and approve/request changes?

**Comment by andresag01** [comment]
@mazimkhan: I checked the latest CI job and it seems that the tls-client examples are failing because the job script is not calling `mbed deploy` like the previous one did ((take a look here)[https://jenkins-internal.mbed.com/job/mbedtls-mbed-os-examples-pr/50/execution/node/15/log/]). However, I recall that the previous Cam-CI job did. Could you please take a look at the job script and modify it so that it uses `mbed deploy`? Thanks!

**Comment by JanneKiiskila** [comment]
For #43 - we can't detect automatically, Mbed OS doesn't support that. But one can modify the `mbed_app.json` with `sed` or `jq` to inject whatever one wants, or (like we do in the Cloud Client/mbed-os-example-client) - provide a number of prepared `.json` files for example under `configs` -folder and replace the `mbed_app.json` with those.

#45 - hard to say, I've not tried it. But, I suppose it SHOULD fail. Haven't tried, though.

**Comment by andresag01** [comment]
@JanneKiiskila: Thanks for the quick reply. Given your comment, we can say that this PR fixes #43 and #101.

With regards to #45 it fails if network is not available. I think the whole point of the ticket is just to make it obvious that its not a problem with the example, but that the target used does not support this example. I think this can be addressed separately though.

**Comment by mazimkhan** [comment]
Added ```mbed deploy``` however a new version of htrun has broken the examples testing. I will fix htrun and get it updated in the cloud. It might take today as htrun is released by Austin.

**Comment by andresag01** [comment]
@mazimkhan: Thanks for the change. I will wait for htrun to be fixed before running the CI again on this PR.

**Comment by mazimkhan** [comment]
Fortunately changing a command line option sufficed the change. CI is ready.

**Comment by JanneKiiskila** [comment]
I think the compilation error is quite self-explanatory, so I wonder what you had in mind - how else to tell? Of course one could add a FAQ section to the `README.md` , but I think most of our target audience should understand that as is.

**Comment by JanneKiiskila** [comment]
Still says; 

```Compile error

Error: easy-connect.h@35,0: #5: cannot open source input file “EthernetInterface.h”: No such file or directory
```

The `mbed deploy` needs to be still pushed in to Oulu CI?

**Comment by mazimkhan** [comment]
@JanneKiiskila ```mbed deploy``` is there. For some targets EthernetInterface.h is present and for some it is not. Is it part of target?

**Comment by andresag01** [comment]
@mazimkhan, @JanneKiiskila: I checked the latest job in more detail. It seems there are a few problems:

* The `mbed deploy` is not missing. The problem with the inclusion of EthernetInterface.h is that the CI is tryings to run tls-client on NCS36510 and KW24D, but neither of these have ethernet ports. So the failure is expected.
* There are multiple errors related to running other examples NCS36510 and KW24D. These errors are unrelated to this PR.
* The following errors keeps popping up, but this seems to be only some problem with the CI:
```
10:48:46 Process Process-1:
10:48:46 Traceback (most recent call last):
10:48:46   File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
10:48:46     self.run()
10:48:46   File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
10:48:46     self._target(*self._args, **self._kwargs)
10:48:46   File "/usr/local/lib/python2.7/dist-packages/mbed_host_tests/host_tests_conn_proxy/conn_proxy.py", line 176, in conn_process
10:48:46     if not connector.write("mbed" * 10, log=True):
10:48:46   File "/usr/local/lib/python2.7/dist-packages/mbed_host_tests/host_tests_conn_proxy/conn_primitive_remote.py", line 133, in write
10:48:46     self.selected_resource.write(payload)
10:48:46   File "/usr/local/lib/python2.7/dist-packages/raas_client/resources/Base.py", line 546, in write
10:48:46     format(Resource.CHANNEL_DEFAULT))
10:48:46 ResourceError: Attempted write when serial0 channel not open!
```

@mazimkhan: When were the NCS36510 and KW24D targets added? I dont recall these being there. Also, what is the other python issue with the CI.

**Comment by mazimkhan** [comment]
@andresag01 Thanks for analysing the logs. There are some known issues with CI. The issues that you pointed out. We are missing a feature to filter platforms based on HW requirements. I will try add that. 
From what is passing I can say the PR is good.

**Comment by andresag01** [comment]
@mazimkhan: Thanks! Could you please let us know when the CI issues are resolved so that we can trigger another build and hopefully merge these changes?

**Comment by mazimkhan** [comment]
@andresag01 some of the issues will be fixed as enhancements and some are ongoing RAAS issues. I would not wait for them to be resolved. As they are yet to be planned and worked upon.

**Comment by andresag01** [comment]
This PR has been approved several reviewers and the relevant tests pass in the CI. Merging.
